File: default-implementation/observer-pattern.rb
class SomeInterestingClass
	def initialize()
		@observers = []
	end
	
	def add_observer o
		@observers.push o
	end
	
	def notify_observers event
		@observers.each { |o| o.do_something event }
	end
	
	def action_A
		notify_observers 'action A'
	end
	
	def action_B
		notify_observers 'action B'
	end
	
	def work
		10.times do
			puts "Im interesting and Im gonna do something..."
			
			if Random.new().rand(2) == 0 then
				action_A()
			else
				action_B()
			end
				
			puts
		end
	end
end

class SomeCuriousClass
	def initialize name
		@name = name
	end

	def do_something event
		puts "observer #{@name} handled event #{event}"
	end
end

interestingThing = SomeInterestingClass.new

interestingThing.add_observer SomeCuriousClass.new "one"
interestingThing.add_observer SomeCuriousClass.new "two"
interestingThing.add_observer SomeCuriousClass.new "three"

interestingThing.work()

File: default-implementation/observer-pattern.py
from random import randint as rand

class SomeInterestingClass:
	def __init__(self):
		self.observers = []
	
	def addobserver(self, o):
		self.observers.append(o)
	
	def notifyobservers(self, event):
		for o in self.observers:
			o.dosomething(event)
	
	def actionA(self):
		self.notifyobservers('action A')
	
	def actionB(self):
		self.notifyobservers('action B')
			
	def work(self):
		for i in range(10):
			print("Im interesting and Im gonna do something...")
			
			if rand(1, 2) == 1:
				self.actionA()
			else:
				self.actionB()
			
			print("")

class SomeCuriousClass:
	def __init__(self, name):
		self.name = name

	def dosomething(self, event):
		print("observer %s handled event %s" % (self.name, event))

interestingThing = SomeInterestingClass()

interestingThing.addobserver(SomeCuriousClass("three"))
interestingThing.addobserver(SomeCuriousClass("two"))
interestingThing.addobserver(SomeCuriousClass("one"))

interestingThing.work()